File: DayNightSwitch.m
//
//  DayNightSwitch.m
//  DayNightSwitch
//
//  Created by Finn Gaida on 03.09.16.
//  Copyright © 2016 Finn Gaida. All rights reserved.
//

#import "DayNightSwitch.h"

/// some color constants
#define onKnobColor [UIColor colorWithRed: 0.882 green: 0.765 blue: 0.325 alpha: 1];
#define onSubviewColor [UIColor colorWithRed: 0.992 green: 0.875 blue: 0.459 alpha: 1];
#define offKnobColor [UIColor colorWithRed: 0.894 green: 0.902 blue: 0.788 alpha: 1];
#define offSubviewColor [UIColor colorWithRed: 1 green: 1 blue: 1 alpha: 1];
#define offColor [UIColor colorWithRed: 0.235 green: 0.255 blue: 0.271 alpha: 1];
#define offBorderColor [UIColor colorWithRed: 0.11 green: 0.11 blue: 0.11 alpha: 1];
#define onColor [UIColor colorWithRed: 0.627 green: 0.894 blue: 0.98 alpha: 1];
#define onBorderColor [UIColor colorWithRed: 0.533 green: 0.769 blue: 0.843 alpha: 1];


@interface Knob : UIView

/// Visual state of the knob, animates changes
@property (nonatomic) BOOL on;

/// Horizontally expanded state of the knob, animates changes
@property (nonatomic) BOOL expanded;

/// Round subview of the knob
@property (nonatomic) UIView *subview;

/// Circular subviews on the off state `subview`
@property (nonatomic) NSArray<UIView *> *craters;

@end

@implementation Knob

/// Distance from knob to subview circle
- (CGFloat)subviewMargin {
    return self.frame.size.height / 12;
}

/**
 Sets up the `subview` with the craters as well
 
 - returns: the view
 */
- (UIView *)setupSubview {
    
    UIView *v = [[UIView alloc] initWithFrame: CGRectMake([self subviewMargin], [self subviewMargin], self.frame.size.width - [self subviewMargin] * 2, self.frame.size.height - [self subviewMargin] * 2)];
    v.layer.masksToBounds = true;
    v.layer.cornerRadius = v.frame.size.height / 2;
    v.backgroundColor = offSubviewColor;
    
    for (UIView *c in [self setupCraters]) {
        [v addSubview:c];
    }
    
    self.subview = v;
    return v;
}

/**
 Sets up three craters
 
 - returns: array of set up views
 */
- (NSArray *)setupCraters {
    
    // shortcuts
    CGFloat w = self.frame.size.width;
    CGFloat h = self.frame.size.height;
    
    UIView *topLeft = [[UIView alloc] initWithFrame: CGRectMake(0, h * 0.1, w * 0.2, w * 0.2)];
    UIView *topRight = [[UIView alloc] initWithFrame: CGRectMake(w * 0.5, 0, w * 0.3, w * 0.3)];
    UIView *bottom = [[UIView alloc] initWithFrame: CGRectMake(w * 0.4, h * 0.5, w * 0.25, w * 0.25)];
    
    NSArray<UIView *> *all = @[topLeft, topRight, bottom];
    
    for (UIView *v in all) {
        v.backgroundColor = offSubviewColor;
        v.layer.masksToBounds = YES;
        v.layer.cornerRadius = v.frame.size.height / 2;
        
        UIColor *offC = offKnobColor;
        v.layer.borderColor = offC.CGColor;
        v.layer.borderWidth = [self subviewMargin];
    }
    
    self.craters = all;
    return all;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self.on = NO;
    self.expanded = NO;
    self = [super initWithFrame:frame];
    
    self.layer.masksToBounds = YES;
    self.layer.cornerRadius = self.frame.size.height / 2;
    self.backgroundColor = offKnobColor;
    
    [self addSubview:[self setupSubview]];
    
    [self addObserver:self forKeyPath:@"on" options:0 context:nil];
    [self addObserver:self forKeyPath:@"expanded" options:0 context:nil];
    
    return self;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context {
    if ([keyPath isEqualToString:@"on"]) {
        [UIView animateWithDuration:0.8 delay:0 usingSpringWithDamping:1 initialSpringVelocity:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
            
            if (self.on) {
                self.backgroundColor = onKnobColor;
                self.subview.backgroundColor = onSubviewColor;
            } else {
                self.backgroundColor = offKnobColor;
                self.subview.backgroundColor = offSubviewColor;
            }
            
            BOOL cache = self.expanded;
            self.expanded = cache;
            
        } completion:nil];
        
        [UIView animateWithDuration:0.4 delay:0 usingSpringWithDamping:1 initialSpringVelocity:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
            
            self.subview.transform = CGAffineTransformMakeRotation(M_PI * ((self.on) ? 0.2 : -0.2));
            
        } completion: nil];
    } else if ([keyPath isEqualToString:@"expanded"]) {
        CGFloat newWidth = self.frame.size.height * (self.expanded ? 1.25 : 1);
        CGFloat x = (self.on) ? self.superview.frame.size.width - newWidth - [(DayNightSwitch *)self.superview knobMargin] : self.frame.origin.x;
        
        [UIView animateWithDuration:0.8 delay:0 usingSpringWithDamping:1 initialSpringVelocity:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
            self.frame = CGRectMake(x, self.frame.origin.y, newWidth, self.frame.size.height);
            self.subview.center = CGPointMake((self.on) ? self.frame.size.width - self.frame.size.height / 2 : self.frame.size.height / 2, self.subview.center.y);
            
            for (UIView *v in self.craters) {
                v.alpha = (self.on) ? 0 : 1;
            }
            
        } completion: nil];
    }
}

@end

@interface DayNightSwitch ()

/// Round white knob
@property (nonatomic) Knob *knob;

@property (nonatomic) BOOL moved;

/// This prevents the tap gesture recognizer from interfering the drag movement
@property (nonatomic) BOOL dragging;

@end

/// A switch inspired by [Dribbble](https://dribbble.com/shots/1909289-Day-Night-Toggle-Button-GIF)a
@implementation DayNightSwitch

/// Width of the darker border of the background
- (CGFloat)borderWidth {
    return self.frame.size.height / 7;
}

/// Distance between border and knob
- (CGFloat)knobMargin {
    return self.frame.size.height / 10;
}

/**
 Sets up the `knob`
 
 - returns: the knob view
 */
- (Knob *)setupKnob {
    
    CGFloat w = self.frame.size.height - [self knobMargin] * 2;
    Knob *v = [[Knob alloc] initWithFrame: CGRectMake([self knobMargin], [self knobMargin], w, w)];
    
    self.knob = v;
    return v;
}


/**
 Sets up the border layers
 
 - returns: array containing both layers
 */
- (NSArray *)setupBorders {
    
    CAShapeLayer *b1 = [CAShapeLayer layer];
    CAShapeLayer *b2 = [CAShapeLayer layer];
    UIBezierPath *path = [UIBezierPath bezierPathWithRoundedRect: CGRectMake(0, 0, self.frame.size.width, self.frame.size.height) cornerRadius: self.frame.size.height / 2];
    
    b1.path = path.CGPath;
    b1.fillColor = [UIColor clearColor].CGColor;
    
    UIColor *onC = onBorderColor;
    b1.strokeColor = onC.CGColor;
    b1.lineWidth = [self borderWidth];
    self.onBorder = b1;
    
    b2.path = path.CGPath;
    b2.fillColor = [UIColor clearColor].CGColor;
    
    UIColor *offC = offBorderColor;
    b2.strokeColor = offC.CGColor;
    b2.lineWidth = [self borderWidth];
    self.offBorder = b2;
    
    return @[b1, b2];
}


/**
 Creates 7 stars with different location and size
 
 - returns: an array of set up views
 */
- (NSArray *)setupStars {
    
    // shortcuts
    CGFloat w = self.frame.size.width;
    CGFloat h = self.frame.size.height;
    
    CGFloat x = h * 0.05;
    UIView *s1 = [[UIView alloc] initWithFrame: CGRectMake(w * 0.5, h * 0.16, x, x)];
    UIView *s2 = [[UIView alloc] initWithFrame: CGRectMake(w * 0.62, h * 0.33, x * 0.6, x * 0.6)];
    UIView *s3 = [[UIView alloc] initWithFrame: CGRectMake(w * 0.7, h * 0.15, x, x)];
    UIView *s4 = [[UIView alloc] initWithFrame: CGRectMake(w * 0.83, h * 0.39, x * 1.4, x * 1.4)];
    UIView *s5 = [[UIView alloc] initWithFrame: CGRectMake(w * 0.7, h * 0.54, x * 0.8, x * 0.8)];
    UIView *s6 = [[UIView alloc] initWithFrame: CGRectMake(w * 0.52, h * 0.73, x * 1.3, x * 1.3)];
    UIView *s7 = [[UIView alloc] initWithFrame: CGRectMake(w * 0.82, h * 0.66, x * 1.1, x * 1.1)];
    
    NSArray *all = @[s1, s2, s3, s4, s5, s6, s7];
    
    for (UIView *s in all) {
        s.layer.masksToBounds = YES;
        s.layer.cornerRadius = s.frame.size.height / 2;
        s.backgroundColor = [UIColor whiteColor];
    }
    
    self.stars = all;
    return all;
}


/**
 Sets up the `cloud`
 
 - returns: the image view
 */
- (UIImageView *)setupCloud {
    
    UIImageView *v = [[UIImageView alloc] initWithFrame: CGRectMake(self.frame.size.width / 3, self.frame.size.height * 0.4, self.frame.size.width / 3, self.frame.size.width * 0.23)];
    v.image = [UIImage imageNamed:@"cloud"];
    v.transform = CGAffineTransformMakeScale(0, 0);
    
    // this should be done with UIBezierPaths...
    
    self.cloud = v;
    return v;
}

// MARK: handling touch events
- (void)proccessTouches:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    if (!self.moved) { self.on = !self.on; return; }
    CGFloat x = [(UITouch *)touches.allObjects.lastObject locationInView:self].x;
    
    if (x > self.frame.size.width / 2 && !self.on) {
        self.on = YES;
    } else if (x < self.frame.size.width / 2 && self.on) {
        self.on = NO;
    }
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    self.dragging = YES;
    self.knob.expanded = YES;
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    self.moved = YES;
    [self proccessTouches:touches withEvent:event];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self proccessTouches:touches withEvent:event];
    self.knob.expanded = NO;
    self.dragging = NO;
    self.moved = NO;
}

- (void)touchesCancelled:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self touchesEnded:touches withEvent:event];
}


// MARK: Initializers
- (instancetype)initWithCenter:(CGPoint)center {
    CGFloat height = 30;
    CGFloat width = height * 1.75;
    
    self = [super initWithFrame: CGRectMake(center.x - width / 2, center.y - height / 2, width, height)];
    [self commonInit];
    
    return self;
}

/**
 Init method called by all initializers. The switch is initialized off by default
 */
- (void)commonInit {
    self.moved = NO;
    self.dragging = NO;
    self.layer.masksToBounds = YES;
    self.layer.cornerRadius = self.frame.size.height / 2;
    self.backgroundColor = [UIColor colorWithRed: 0.235 green: 0.255 blue: 0.271 alpha: 1];
    
    NSArray *borders = [self setupBorders];
    [self.layer addSublayer:borders[0]];
    [self.layer addSublayer:borders[1]];
    
    for (UIView *v in [self setupStars]) {
        [self addSubview:v];
    }
    
    [self addSubview:[self setupKnob]];
    [self addSubview:[self setupCloud]];
    
    [self addObserver:self forKeyPath:@"on" options:0 context:nil];
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    [self commonInit];
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    [self commonInit];
    return self;
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context {
    if ([keyPath isEqualToString:@"on"]) {
        // call the action closure
        if (self.changeAction) { self.changeAction(self.on); }
        
        self.knob.on = self.on;
        
        [UIView animateWithDuration:0.4 delay:0 usingSpringWithDamping:1 initialSpringVelocity:0 options:UIViewAnimationOptionAllowUserInteraction animations:^{
            
            CGFloat knobRadius = self.knob.frame.size.width / 2;
            
            if (self.on) {
                self.knob.center = CGPointMake(self.frame.size.width - knobRadius - [self knobMargin], self.knob.center.y);
                
                self.backgroundColor = onColor;
                self.offBorder.strokeStart = 1.0;
                self.cloud.transform = CGAffineTransformIdentity;
            } else {
                self.knob.center = CGPointMake(knobRadius + [self knobMargin], self.knob.center.y);
                
                self.backgroundColor = offColor;
                self.offBorder.strokeEnd = 1.0;
                self.cloud.transform = CGAffineTransformMakeScale(0, 0);
            }
            
            for (int i = 0; i < self.stars.count; i++) {
                UIView *star = self.stars[i];
                star.alpha = (self.on) ? 0 : 1;
                
                
                dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.1 * i * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                    star.transform = CGAffineTransformMakeScale(1.5, 1.5);
                    
                    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.05 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
                        star.transform = CGAffineTransformIdentity;
                    });
                });
            }
            
        } completion:^(BOOL finished) {
            
            // reset the values
            if (self.on) {
                self.offBorder.strokeStart = 0.0;
                self.offBorder.strokeEnd = 0.0;
            } else {
                self.offBorder.strokeStart = 0.0;
                self.offBorder.strokeEnd = 1.0;
            }
        }];
    }
}

@end

File: DayNightSwitch.swift
//
//  DayNightSwitch
//  DayNightSwitch
//
//  Created by Finn Gaida on 02.09.16.
//  Copyright © 2016 Finn Gaida. All rights reserved.
//

import UIKit

/// View holding the managing of the knob and its subviews
@IBDesignable internal class Knob: UIView {
    
    /// some color constants
    let onKnobColor = UIColor(red: 0.882, green: 0.765, blue: 0.325, alpha: 1)
    let onSubviewColor = UIColor(red: 0.992, green: 0.875, blue: 0.459, alpha: 1)
    let offKnobColor = UIColor(red: 0.894, green: 0.902, blue: 0.788, alpha: 1)
    let offSubviewColor = UIColor(red: 1, green: 1, blue: 1, alpha: 1)
    
    /// Visual state of the knob, animates changes
    @IBInspectable var on: Bool {
        didSet {
            
            UIView.animate(withDuration: 0.8, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: 0, options: UIViewAnimationOptions.allowUserInteraction, animations: { 
                
                self.backgroundColor = self.on ? self.onKnobColor : self.offKnobColor
                
                if let v = self.subview {
                    v.backgroundColor = self.on ? self.onSubviewColor : self.offSubviewColor
                }
                
                }, completion: nil)
            
            
            UIView.animate(withDuration: 0.4, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: 0, options: UIViewAnimationOptions.allowUserInteraction, animations: { 
                
                if let v = self.subview {
                    v.transform = CGAffineTransform(rotationAngle: CGFloat(M_PI) * (self.on ? 0.2 : -0.2))
                }
                
                }, completion: nil)
            
            let cache = self.expanded
            self.expanded = cache
        }
    }
    
    /// Horizontally expanded state of the knob, animates changes
    var expanded: Bool {
        didSet {
            
            guard let sup = self.superview as? DayNightSwitch, let sub = self.subview, let dots = self.craters else { return }
            let newWidth = self.frame.height * (self.expanded ? 1.25 : 1)
            let x = self.on ? sup.frame.width - newWidth - sup.knobMargin : self.frame.origin.x
            
            UIView.animate(withDuration: 0.8, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: 0, options: UIViewAnimationOptions.allowUserInteraction, animations: {
                self.frame = CGRect(x: x, y: self.frame.origin.y, width: newWidth, height: self.frame.height)
                sub.center = CGPoint(x: self.on ? self.frame.width - self.frame.height / 2 : self.frame.height / 2, y: sub.center.y)
                
                dots.forEach({ $0.alpha = self.on ? 0 : 1 })
                
                }, completion: nil)
        }
    }
    
    /// Distance from knob to subview circle
    var subviewMargin: CGFloat {
        get {
            return self.frame.height / 12
        }
    }
    
    /// Round subview of the knob
    var subview: UIView?
    
    /**
     Sets up the `subview` with the craters as well
     
     - returns: the view
     */
    func setupSubview() -> UIView {
    
        let v = UIView(frame: CGRect(x: subviewMargin, y: subviewMargin, width: self.frame.width - subviewMargin * 2, height: self.frame.height - subviewMargin * 2))
        v.layer.masksToBounds = true
        v.layer.cornerRadius = v.frame.height / 2
        v.backgroundColor = offSubviewColor
        
        setupCraters().forEach { v.addSubview($0) }
        
        subview = v
        return v
    }
    
    /// Circular subviews on the off state `subview`
    var craters: [UIView]?
    
    /**
     Sets up three craters
     
     - returns: array of set up views
     */
    func setupCraters() -> [UIView] {
        
        // shortcuts
        let w = self.frame.width
        let h = self.frame.height
        
        let topLeft = UIView(frame: CGRect(x: 0, y: h * 0.1, width: w * 0.2, height: w * 0.2))
        let topRight = UIView(frame: CGRect(x: w * 0.5, y: 0, width: w * 0.3, height: w * 0.3))
        let bottom = UIView(frame: CGRect(x: w * 0.4, y: h * 0.5, width: w * 0.25, height: w * 0.25))
        
        let all = [topLeft, topRight, bottom]
        all.forEach { (v) in
            v.backgroundColor = offSubviewColor
            v.layer.masksToBounds = true
            v.layer.cornerRadius = v.frame.height / 2
            v.layer.borderColor = offKnobColor.cgColor
            v.layer.borderWidth = subviewMargin
        }
        
        craters = all
        return all
    }
    
    override init(frame: CGRect) {
        self.on = false
        self.expanded = false
        super.init(frame: frame)
        
        self.layer.masksToBounds = true
        self.layer.cornerRadius = self.frame.height / 2
        self.backgroundColor = offKnobColor
        
        self.addSubview(setupSubview())
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}

/// A switch inspired by [Dribbble](https://dribbble.com/shots/1909289-Day-Night-Toggle-Button-GIF)
@IBDesignable open class DayNightSwitch: UIView {
    
    // some constant colors
    let offColor = UIColor(red: 0.235, green: 0.255, blue: 0.271, alpha: 1)
    let offBorderColor = UIColor(red: 0.11, green: 0.11, blue: 0.11, alpha: 1)
    let onColor = UIColor(red: 0.627, green: 0.894, blue: 0.98, alpha: 1)
    let onBorderColor = UIColor(red: 0.533, green: 0.769, blue: 0.843, alpha: 1)

    /// Width of the darker border of the background
    open var borderWidth: CGFloat {
        get {
            return self.frame.height / 7
        }
    }
    
    /// Distance between border and knob
    open var knobMargin: CGFloat {
        get {
            return self.frame.height / 10
        }
    }
    
    /// Called as soon as the value changes (probably because the user tapped it)
    open var changeAction: ((Bool) -> ())?
    
    /// Determines the state of the button, animates changes
    @IBInspectable open var on: Bool = true {
        didSet {
            
            // call the action closure
            if let c = self.changeAction { c(self.on) }
            
            if let k = self.knob {
                k.on = on
                
                UIView.animate(withDuration: 0.4, delay: 0, usingSpringWithDamping: 1, initialSpringVelocity: 0, options: UIViewAnimationOptions.allowUserInteraction, animations: {
                    
                    let knobRadius = k.frame.width / 2
                    k.center = CGPoint(x: (self.on ? self.frame.width - knobRadius - self.knobMargin : knobRadius + self.knobMargin), y: k.center.y)
                    
                    self.backgroundColor = self.on ? self.onColor : self.offColor
                    
                    if let b = self.offBorder {
                        if self.on {
                            b.strokeStart = 1.0
                        } else {
                            b.strokeEnd = 1.0
                        }
                    }
                    
                    if let all = self.stars {
                        all.enumerated().forEach({ (i, star) in
                            star.alpha = self.on ? 0 : 1
                            
                            // let them shine
                            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + Double(Int64(0.1 * Double(i) * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC), execute: {
                                star.transform = CGAffineTransform(scaleX: 1.5, y: 1.5)
                                
                                DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + Double(Int64(0.05 * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC), execute: {
                                    star.transform = CGAffineTransform.identity
                                })
                            })
                        })
                    }
                    
                    if let c = self.cloud {
                        c.transform = self.on ? CGAffineTransform.identity : CGAffineTransform(scaleX: 0, y: 0)
                    }
                    
                    }, completion: { _ in
                
                        // reset the values
                        if let b = self.offBorder {
                            if self.on {
                                b.strokeStart = 0.0
                                b.strokeEnd = 0.0
                            } else {
                                b.strokeStart = 0.0
                                b.strokeEnd = 1.0
                            }
                        }
                })
            }
        }
    }
    
    /// Round white knob
    var knob: Knob?
    
    /**
     Sets up the `knob`
     
     - returns: the knob view
     */
    func setupKnob() -> Knob {
        
        let w = self.frame.height - knobMargin * 2
        let v = Knob(frame: CGRect(x: knobMargin, y: knobMargin, width: w, height: w))
        knob = v
        return v
    }
    
    /// Dark blue border layer
    open var offBorder: CAShapeLayer?
    
    /// Light blue layer below the `offBorder`
    open var onBorder: CAShapeLayer?
    
    /**
     Sets up the border layers
     
     - returns: tuple containing both layers
     */
    func setupBorders() -> (CAShapeLayer, CAShapeLayer) {
        
        let b1 = CAShapeLayer()
        let b2 = CAShapeLayer()
        let path = UIBezierPath(roundedRect: CGRect(x: 0, y: 0, width: self.frame.width, height: self.frame.height), cornerRadius: self.frame.height / 2)

        b1.path = path.cgPath
        b1.fillColor = UIColor.clear.cgColor
        b1.strokeColor = onBorderColor.cgColor
        b1.lineWidth = borderWidth
        onBorder = b1
        
        b2.path = path.cgPath
        b2.fillColor = UIColor.clear.cgColor
        b2.strokeColor = offBorderColor.cgColor
        b2.lineWidth = borderWidth
        offBorder = b2
        
        return (b1, b2)
    }
    
    /// Small white dots on the off state background
    open var stars: [UIView]?
    
    /**
     Creates 7 stars with different location and size
     
     - returns: an array of set up views
     */
    func setupStars() -> [UIView] {
        
        // shortcuts
        let w = self.frame.width
        let h = self.frame.height
        
        let x = h * 0.05
        let s1 = UIView(frame: CGRect(x: w * 0.5, y: h * 0.16, width: x, height: x))
        let s2 = UIView(frame: CGRect(x: w * 0.62, y: h * 0.33, width: x * 0.6, height: x * 0.6))
        let s3 = UIView(frame: CGRect(x: w * 0.7, y: h * 0.15, width: x, height: x))
        let s4 = UIView(frame: CGRect(x: w * 0.83, y: h * 0.39, width: x * 1.4, height: x * 1.4))
        let s5 = UIView(frame: CGRect(x: w * 0.7, y: h * 0.54, width: x * 0.8, height: x * 0.8))
        let s6 = UIView(frame: CGRect(x: w * 0.52, y: h * 0.73, width: x * 1.3, height: x * 1.3))
        let s7 = UIView(frame: CGRect(x: w * 0.82, y: h * 0.66, width: x * 1.1, height: x * 1.1))
        
        let all = [s1, s2, s3, s4, s5, s6, s7]
        all.forEach { (s) in
            s.layer.masksToBounds = true
            s.layer.cornerRadius = s.frame.height / 2
            s.backgroundColor = UIColor.white
        }
        
        stars = all
        return all
    }
    
    /// Cloud image visible on top of the on state knob
    open var cloud: UIImageView?
    
    /**
     Sets up the `cloud`
     
     - returns: the image view
     */
    func setupCloud() -> UIImageView {
        
        let v = UIImageView(frame: CGRect(x: self.frame.width / 3, y: self.frame.height * 0.4, width: self.frame.width / 3, height: self.frame.width * 0.23))
        v.image = UIImage(named: "cloud")
        
        v.transform = CGAffineTransform(scaleX: 0, y: 0)

        // this should be done with UIBezierPaths...
        
        cloud = v
        return v
    }
    
    // MARK: handling touch events
    func proccess(_ touches: Set<UITouch>, withEvent event: UIEvent?) {
        if !moved { return self.on = !self.on }
        guard let x = touches.first?.location(in: self).x else { return }
        
        if x > self.frame.width / 2 && !self.on {
            self.on = true
        } else if x < self.frame.width / 2 && self.on {
            self.on = false
        }
    }

    /// This prevents the tap gesture recognizer from interfering the drag movement
    var dragging: Bool = false
    override open func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        dragging = true
        if let k = self.knob { k.expanded = true }
    }
    
    var moved: Bool = false
    override open func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        moved = true
        proccess(touches, withEvent: event)
    }
    
    override open func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.proccess(touches, withEvent: event)
        if let k = self.knob { k.expanded = false }
        dragging = false
        moved = false
    }
    
    override open func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.touchesEnded(touches , with: event)
    }
    
    
    // MARK: Initializers
    public init(center: CGPoint) {
        let height: CGFloat = 30
        let width: CGFloat = height * 1.75
        super.init(frame: CGRect(x: center.x - width / 2, y: center.y - height / 2, width: width, height: height))
        commonInit()
    }
    
    /**
     Init method called by all initializers. The switch is initialized off by default
     */
    func commonInit() {
        self.layer.masksToBounds = true
        self.layer.cornerRadius = self.frame.height / 2
        self.backgroundColor = UIColor(red: 0.235, green: 0.255, blue: 0.271, alpha: 1)

        self.layer.addSublayer(setupBorders().0)
        self.layer.addSublayer(setupBorders().1)
        setupStars().forEach { self.addSubview($0) }
        self.addSubview(setupKnob())
        self.addSubview(setupCloud())
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }
    
    override public init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
}
